extensions/amp-analytics/0.1/linker.js

```diff
@@ -0,0 +1,179 @@
+    });
+  }
+
+
+  /**
+   * Go through key value pairs and resolve any macros that may exist.
+   * @param {(Object<string, string>|undefined)} pairs
+   * @return {Promise<string>}
+   */
+  resolveMacros_(pairs) {
+    const expandedPairs = {};
+    const expansionPromises = [];
+
+    for (const key in pairs) {
+      // TODO(ccordry): change this to call new expander once fully launched.
+      const promise = this.urlReplacementService_.expandStringAsync(pairs[key]);
+      // Add value promises into array so that we can keep track of resolution.
+      // Also add to new obj, so that we can still keep track of key-value
+      // pairs.
+      expansionPromises.push(promise);
+      expandedPairs[key] = promise;
+    }
+
+    return Promise.all(expansionPromises).then(() => {
+      return expandedPairs;
+    });
+  }
+
+
+  /**
+   * Generate the completed querystring.
+   * <paramName>=<version>~<checksum>~<key1>~<value1>~<key2>~<value2>...
+   * @param {string} version
+   * @param {?Object<string, string>} expandedPairs
+   * @return {string}
+   */
+  generateParam_(version, expandedPairs) {
+    const encodedPairs = this.encodePairs_(expandedPairs);
+    const checksum = this.getCheckSum_(encodedPairs);
+    return version + DELIMITER + checksum + encodedPairs;
+  }
+
+
+  /**
+   * Create a unique checksum hashing the fingerprint and a few other values.
+   * base64(CRC32(fingerprint + timestampRoundedInMin + kv pairs))
+   * @param {?Object<string, string>} encodedPairs
+   * @return {string}
+   */
+  getCheckSum_(encodedPairs) {
+    const fingerprint = this.getFingerprint_();
+    const timestamp = this.getRoundedTimestamp_();
+    const goo = crc32String(fingerprint + timestamp + encodedPairs);
+    return b64EncodeUnicode(goo);
+  }
+
+
+  /**
+   * Generates a semi-unique value for page visitor.
+   * User Agent + timezone + language.
+   * @return {string}
+   */
+  getFingerprint_() {
+    const date = new Date();
+    const timezone = date.getTimezoneOffset();
+
+    return navigator.userAgent + timezone + navigator.language;
+  }
+
+  /**
+   * After macros have resolved, encode their values and join them together.
+   * @param {?Object<string, string>} expandedPairs
+   * @return {string}
+   */
+  encodePairs_(expandedPairs) {
+    const keys = Object.keys(expandedPairs);
+    if (!keys.length) {
+      return '';
+    }
+
+    let result = DELIMITER;
+
+    keys.forEach(key => {
+      const encodedKey = b64EncodeUnicode(key);
+      const encodedVal = b64EncodeUnicode(expandedPairs[key]);
+      result += encodedKey + DELIMITER + encodedVal;
+    });
+
+    return result;
+  }
+
+
+  /**
+   * Rounded time used to check if t2 - t1 is within our time tolerance.
+   * @return {number}
+   */
+  getRoundedTimestamp_() {
+    return Math.round(Date.now() / TIME_TOLERANCE);
+  }
+}
+
+
+/**
+ * Escape the given string with UTF-8 then base64 encode it.
+ * @param {string} str
+ */
+function b64EncodeUnicode(str) {
```

Comment:
I copied the example from MDN here https://developer.mozilla.org/en-US/docs/Web/API/WindowBase64/Base64_encoding_and_decoding. Even thought we already have a similar thing in AMP, I thought it might be easier to use this easily accessible version since other people will have to decode this value server side. Happy to change if you have a better idea.

Reply:
so the extra complexity is to address Unicode issue. Can you compare this method with existing `base64UrlEncodeFromBytes` method? To see if they generate same results for non-Unicode strings. If so, we can take it as an enhancement to the existing method.